### PR TITLE
Fix bug where duplicate PSBT keys are accepted

### PR DESCRIPTION
**Pull request by erasmospunk**
As per the BIP 174 spec a PSBT key cannot be duplicated,
however the current code accepts key duplication.
The PSBT key/value entries can be duplicated when the value
is `empty()` or `IsNull()` for `CScript` or `CTxOut` respectively
and if those key/value entries are serialized before the non-empty ones.

For example, the following PSBT, included in the test vectors,
contains a duplicate field:

```
// magic
70736274ff

// global tx
//// key
0100
//// value
2a02000000000140420f000000000017a9146e91b72d5593e7d4391e2ff44e91e985c31641f08700000000
//// separator
00

// no inputs

// outputs
//// key PSBT_OUT_WITNESSSCRIPT
0101
//// value (empty script)
00
//// key PSBT_OUT_WITNESSSCRIPT (same as the above)
0101
//// value (an OP_RETURN script)
016a
//// separator
00
```

### CODE REVIEW

File: src/psbt.h [hunk 1]
```diff
@@ -243,7 +246,7 @@ struct PSBTInput
                 }
                 // Unknown stuff
                 default:
-                    if (unknown.count(key) > 0) {
+                    if (!key_lookup.emplace(key).second) {
```
Comment (achow101) [inline src/psbt.h]: I don't think this check for an unknown key is necessary in the same way the it is not needed for partial sigs or hd keypaths. It is a map of unknowns and we aren't checking if something is null.

Comment (erasmospunk) [reply src/psbt.h]: You are correct, I left it for visual consistency (probably should have done the same for `PSBT_IN_PARTIAL_SIG`).

If you feel strong about it, will revert it for the unknowns.

**Comment by promag** [comment]
Concept ACK, could include more invalid tests?

File: src/psbt.h (hunk 1 above)
Comment (erasmospunk) [reply src/psbt.h]: Finally have reverted it for the unknowns.

**Comment by erasmospunk** [comment]
Rebased to the latest master and added some more test cases to the rpc_psbt.json as @promag suggested.

**Comment by instagibbs** [comment]
concept ACK

**Comment by erasmospunk** [comment]
Sorry, added the changes as a separate commit instead of amending, for easier review.

**Comment by achow101** [comment]
ACK b505d774eda28cbe965cd9d9ded3c9a83b9392fa, but needs squashing

Checked the individual test cases by hand and verified they contain duplicate keys with values that meet the previous empty conditions. Also checked that those tests fail on master.

**Comment by erasmospunk** [comment]
@achow101 squashed

**Comment by achow101** [comment]
ACK 9743432034586385cfef87df4b377c255ed0cba8

**Comment by instagibbs** [comment]
code review ACK https://github.com/bitcoin/bitcoin/pull/16821/commits/9743432034586385cfef87df4b377c255ed0cba8

did not verify the test cases

**Pull request merged**